### ISSUE TITLE
Accessing Tango from your macros

### ISSUE THREAD

**Issue by dschick** [sardana-org/sardana#1453]
This is a rather general question about the "correct" way to access Tango devices in Sardana macros.
We do have the generic case of module that holds serveral macro functions which all interact with the same `tango.DeviceProxy`.
Currently we create that proxy within every function, e.g.
```python
from sardana.macroserver.macro import Macro, macro, Type
from tango import DeviceProxy

@macro()
def laser_on(self):
    """Macro laser_on"""
    proxy = DeviceProxy('laser/ThorlabsShutter/0')
    proxy.Open()
    self.output("Laser shutter open!")

@macro()
def laser_off(self):
    """Macro laser_off"""
    proxy = DeviceProxy('laser/ThorlabsShutter/0')
    proxy.Close()
    self.output("Laser shutter closed!")

...
```
Is this a smart way of doing it?
Or could I open the proxy ones and use the same proxy object by all macro functions?

**Comment by tiagocoutinho** [comment]
Hi @dschick,

First let me say that there is nothing wrong with your code. It is not inefficient as your intuition might be telling you.

### TLDR;

You can use a helper that caches DeviceProxy objects:
```python
import functools
import tango
from sardana.macroserver.macro import Macro, macro

Device = functools.lru_cache(maxsize=1024)(tango.DeviceProxy)

@macro()
def laser_on(self):
    """Macro laser_on"""
    proxy = Device('laser/ThorlabsShutter/0')
    proxy.Open()
    self.output("Laser shutter open!")
```

### long explanation

As I said above, IMHO, there is no problem with your code. This is mainly because tango/corba cache tcp connections for a period of up to 30s of inactivity (i am not absolutely sure about this time. We might need to check C++ code to confirm).

So, the second time you invoke one of your methods (if the time between them is short enough) the connection to the server is already established. From the sniff on the network I still see there is an extra packet but this is neglectable when compared with the time to establish a connection. If want to prevent this extra packet (and I think you should) the snippet I wrote above should help.

The downside is, if you don't call methods too often and since the communication is closed after a period of inactivity, 
the jitter is enormous (if you care about this sort of thing) and I don't thing there is anything you can do about it.

**Comment by reszelaz** [comment]
Hi all,

This is a very interesting discussion. Thanks @dschick for raising it and @tiagocoutinho for sharing your example and description. 

We had it already many times at ALBA. Actually our scientists who develop macros always _complain_ there are too many ways of accessing Tango and they would like to have a sort of guide on which way to choose. To fullfill their requirement I have proposed this [FAQ](https://sardana-controls.org/users/faq.html#how-to-access-tango-from-within-macros-or-controllers) in the docs. It points to different parts of the documentation which was also recently added (see https://github.com/sardana-org/sardana/pull/1341).

Feel free to comment, propose changes and/or enhance to this documentation.

Many thanks!

**Comment by marouanebj** [comment]
Hello,
just for my information is it correct to do this ?

```python
from taurus import Device # added

# import functools # removed
import tango
from sardana.macroserver.macro import Macro, macro

# Device = functools.lru_cache(maxsize=1024)(tango.DeviceProxy) # removed

@macro()
def laser_on(self):
    """Macro laser_on"""
    proxy = Device('laser/ThorlabsShutter/0')
    proxy.Open()
    self.output("Laser shutter open!")
```

thanks

**Comment by dschick** [comment]
thanks @tiagocoutinho for the detailed explaination.
Since I do not care so much about latency, I am fine with the tango/corba tcp cache.
This was actually my worry, that I will have multiple tcp connections running at the same time ...

Also thanks @reszelaz for the clarification about the docs.
I carefully read it, but it was not tackeling my issue.
Maybe it is worth adding the info of @tiagocoutinho to it:
>As I said above, IMHO, there is no problem with your code. This is mainly because tango/corba cache tcp connections for a period of up to 30s of inactivity (i am not absolutely sure about this time. We might need to check C++ code to confirm).

**Comment by reszelaz** [comment]
> Maybe it is worth adding the info of @tiagocoutinho to it:

Good idea!
A question to all, what do you think about adding this documentation to Tango / PyTango? Depending where it fits better. Or maybe it is already documented somewhere?
If it is not, I think it may be interesting to add it there cause it may be usefyl not only to the Sardana users.
If you agree then we could link it from our documentatation using the inter-sphinx  links.
If you don't agree, I don't mind adding it directly to Sardana as well.

Thanks again! I appreciate all your efforts in improving the docs!

**Comment by bourtemb** [comment]
Tango is using CORBA underneath, omniORB in the case of pyTango and cppTango.
omniORB has a thread which takes care of closing the idle connections.
This is described in omniORB documentation: http://omniorb.sourceforge.net/omni42/omniORB/omniORB006.html#sec%3AconnShutdown
This is configurable in omniORB using some omniORB configuration parameters (Default values are described here: http://omniorb.sourceforge.net/omni42/omniORB/omniORB004.html)

> outConScanPeriod    default = 120 s
> Idle timeout in seconds for outgoing (i.e. client initiated) connections. If a connection has been idle for this amount of time, the ORB closes it.
> 
> inConScanPeriod    default = 180 s
> Idle timeout in seconds for incoming connections. If a connection has been idle for this amount of time, the ORB closes it.
> 
> scanGranularity    default = 5 s
> As explained in chapter 6, omniORB regularly scans incoming and outgoing connections, so it can close unused ones. This value is the granularity in seconds at which the ORB performs its scans. A value of zero turns off the scanning altogether.
